src/shim/gate.cc

```diff
@@ -0,0 +1,45 @@
+#include "gate.h"
+
+#include <cstring>
+#include <unistd.h>
+
+#include <sched.h>
+
+// (rwails) The number times we should increment the counter and 
+// check the atomic bool before we fall back to the semaphore.
+// TODO: Move to an environment variable?
+#define SHD_GATE_SPIN_MAX 8096
+
+void gate_init(Gate *gate) {
+  std::memset(gate, 0, sizeof(Gate));
```

Comment:
IIRC memset'ing an object to zero is a bit of a gray area for type safety in the standard. It's *probably* fine in practice, but better to either use struct initialization to default-init the struct in a type-aware way (`*gate = {0}`), and/or explicitly init all members (`gate->x` is currently not initialized other than in this memset).

Reply:
Fixed. n.b. C++ doesn't support the lovely `= {0}` syntax to zero all struct members.